runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java

```diff
@@ -295,13 +301,43 @@ public void initializeState(StateInitializationContext context) throws Exception
 
     sideInputReader = NullSideInputReader.of(sideInputs);
 
-    // maybe init by initializeState
-    if (nonKeyedStateInternals == null) {
-      if (keyCoder != null) {
-        nonKeyedStateInternals =
-            new FlinkKeyGroupStateInternals<>(keyCoder, getKeyedStateBackend());
-      } else {
-        nonKeyedStateInternals = new FlinkSplitStateInternals<>(getOperatorStateBackend());
+    if (keyCoder != null) {
+      nonKeyedStateInternals = new FlinkKeyGroupStateInternals<>(keyCoder, getKeyedStateBackend());
+    } else {
+      nonKeyedStateInternals = new FlinkSplitStateInternals<>(getOperatorStateBackend());
+    }
+
+    if (getKeyedStateBackend() != null) {
+      int numberOfKeyGroups = getKeyedStateBackend().getNumberOfKeyGroups();
+      KeyGroupsList localKeyGroupRange = getKeyedStateBackend().getKeyGroupRange();
+
+      for (KeyGroupStatePartitionStreamProvider streamProvider : context.getRawKeyedStateInputs()) {
+        DataInputViewStreamWrapper div = new DataInputViewStreamWrapper(streamProvider.getStream());
+
+        int keyGroupIdx = streamProvider.getKeyGroupId();
+        checkArgument(
+            localKeyGroupRange.contains(keyGroupIdx),
+            "Key Group " + keyGroupIdx + " does not belong to the local range.");
+
+        // TODO We read the timers here just to progress the stream in order to restore
+        // FlinkKeyGroupStateInternals below. The timer service is automatically initialized
+        // before calling the initializeState method.
+        InternalTimerServiceSerializationProxy serializationProxy =
+            new InternalTimerServiceSerializationProxy(
+                new HashMap(),
+                getUserCodeClassloader(),
+                numberOfKeyGroups,
+                localKeyGroupRange,
+                this,
+                getProcessingTimeService(),
+                keyGroupIdx);
+        serializationProxy.read(streamProvider.getStream());
```

Comment:
@aljoscha This is the hack I use to make it work for Flink 1.5. I think it would be nice for Flink to partially consume the input stream and then pass on the remainder to the user code. Is there any way to do that without hacks like the above?